src/main/kotlin/com/exactpro/th2/conn/dirty/fix/brokenconn/strategy/StrategyState.kt

```diff
@@ -15,38 +15,52 @@
  */
 package com.exactpro.th2.conn.dirty.fix.brokenconn.strategy
 
+import com.exactpro.th2.common.grpc.MessageID
 import com.exactpro.th2.conn.dirty.fix.brokenconn.configuration.RuleConfiguration
+import com.google.protobuf.TextFormat.shortDebugString
 import io.netty.buffer.ByteBuf
 import io.netty.buffer.CompositeByteBuf
 import io.netty.buffer.Unpooled
 import java.time.Instant
+import java.util.Collections
 import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.atomic.AtomicInteger
 import java.util.concurrent.atomic.AtomicLong
 import java.util.concurrent.locks.ReentrantLock
 import kotlin.concurrent.withLock
+import mu.KotlinLogging
 
 class StrategyState(val config: RuleConfiguration? = null) {
     val startTime: Instant = Instant.now()
     val type = config?.ruleType ?: RuleType.DEFAULT
     val batchMessageCache: CompositeByteBuf = Unpooled.compositeBuffer()
+    val messageIDs = Collections.synchronizedList(ArrayList<MessageID>())
+
 
     private val writeLock = ReentrantLock()
     private val missedMessagesCache: MutableMap<Int, ByteBuf> = ConcurrentHashMap<Int, ByteBuf>()
 
-    val batchMessageCacheSize = AtomicInteger(0)
+    private val batchMessageCacheSize = AtomicInteger(0)
```

Comment:
Maybe avoid using Atomic* classes and execute all operation under lock, because you change state of atomic variable as a part of complex operation